react-components/src/components/Reveal3DResources/queryMappedData.ts

```diff
@@ -0,0 +1,169 @@
+            space: fdmConfig.global3dSpace,
+            externalId: `model_3d_${model.modelId}`
+          }
+        }
+      },
+      {
+        equals: {
+          property: [
+            fdmConfig.source.space,
+            `${fdmConfig.source.externalId}/${fdmConfig.source.version}`,
+            'revisionId'
+          ],
+          value: model.revisionId
+        }
+      },
+      {
+        in: {
+          property: [
+            fdmConfig.source.space,
+            `${fdmConfig.source.externalId}/${fdmConfig.source.version}`,
+            'revisionNodeId'
+          ],
+          values: ancestorIds
+        }
+      }
+    ]
+  };
+
+  return await fdmClient.filterAllInstances(filter, 'edge', fdmConfig.source);
+}
+
+async function inspectNode(
+  fdmClient: FdmSDK,
+  dataNode: DmsUniqueIdentifier
+): Promise<InspectResultList> {
+  const inspectionResult = await fdmClient.inspectInstances({
+    inspectionOperations: { involvedViewsAndContainers: {} },
+    items: [
+      {
+        instanceType: 'node',
+        externalId: dataNode.externalId,
+        space: dataNode.space
+      }
+    ]
+  });
+
+  return inspectionResult;
+}
+
+async function filterNodeData<NodeType>(
+  fdmClient: FdmSDK,
+  dataNode: DmsUniqueIdentifier,
+  dataView: Source
+): Promise<NodeType | undefined> {
+  if (dataView === undefined) {
+    return undefined;
+  }
+
+  const dataQueryResult = await fdmClient.filterAllInstances(
+    {
+      and: [
+        { equals: { property: ['node', 'space'], value: dataNode.space } },
+        {
+          equals: {
+            property: ['node', 'externalId'],
+            value: dataNode.externalId
+          }
+        }
+      ]
+    },
+    'node',
+    dataView
+  );
+
+  return dataQueryResult.edges[0]?.properties[dataView.space]?.[
+    `${dataView.externalId}/${dataView.version}`
+  ];
+}
+
+export async function queryMappedData<NodeType>(
+  viewer: Cognite3DViewer,
+  cdfClient: CogniteClient,
+  fdmClient: FdmSDK,
+  fdmConfig: FdmAssetMappingsConfig,
+  clickEvent: PointerEventData
+): Promise<NodeDataResult<NodeType> | undefined> {
+  const intersection = await viewer.getIntersectionFromPixel(
+    clickEvent.offsetX,
+    clickEvent.offsetY
+  );
+
+  if (intersection === null || intersection.type !== 'cad') {
+    return;
+  }
+
+  const cadIntersection = intersection;
+  const model = cadIntersection.model;
+
+  const ancestorIds = await getAncestorNodeIdsForTreeIndex(
+    cdfClient,
+    model,
+    cadIntersection.treeIndex
+  );
+
+  const mappings = await getMappingEdges(fdmClient, fdmConfig, model, ancestorIds);
+
+  if (mappings.edges.length === 0) {
+    return;
+  }
+
+  const dataNode = mappings.edges[0].startNode;
+
+  const inspectionResult = await inspectNode(fdmClient, dataNode);
+
+  const dataView =
+    inspectionResult.items[0]?.inspectionResults.involvedViewsAndContainers?.views[0];
+
+  const nodeData = await filterNodeData<NodeType>(fdmClient, dataNode, dataView);
```

Comment:
Why do we need these 2 requests? I think it should be nice just to return an externalId of connected equipment from `mappings` without anything else. Then the user should define what to do with that. Also, it is quite slow when having so much queries.

Reply:
Right... Perhaps that is the best, yes :thinking: Will ask Deep about that